File: app/src/main/java/com/example/review/New/LibraryStruct.java
package com.example.review.New;

import android.support.annotation.NonNull;

import com.example.review.DataStructureFile.ReviewData;

import java.io.*;

public class LibraryStruct extends StoreData {
    private String text  = "";
    private int    type  = 0;
    public  int    refer = 0;
    public  int    id    = 0;

    public LibraryStruct() {
    }

    public LibraryStruct(LibraryStruct ls) {
        this.text = ls.text;
        this.type = ls.type;
    }

    public void copyOf(LibraryStruct ls) {
        //是引用的话，则不能拷贝
        if (refer > 0) return;

        //拷贝数据
        this.text = ls.text;
        this.type = ls.type;
    }

    public LibraryStruct(String text, int type) {
        this.text = text;
        this.type = type;
    }

    public LibraryStruct(byte[] rawBytes) {
        loadWith(rawBytes);
    }

    public void setIdAuto() {
        if (ReviewData.mLibraries != null) {
            int maxId = 0;
            for (LibraryStruct ls : ReviewData.mLibraries) if (ls.id > maxId) maxId = ls.id;
            id = maxId + 1;
        }
    }

    public String getText() {
        if (refer > 0 && ReviewData.mLibraries != null) {
            for (LibraryStruct ls : ReviewData.mLibraries)
                if (ls.id == refer) return ls.getText();
            return "**引用丢失！**";
        }
        return text;
    }

    public void setText(String text) {
        if (refer > 0) this.text = "";
        else this.text = text;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    @Override
    public void toBytes(DataOutputStream dos) throws IOException {
        dos.writeUTF(text);
        dos.writeInt(type);

        dos.writeInt(id);
        dos.writeInt(refer);
    }

    @Override
    public void loadWith(DataInputStream dis) throws IOException {
        text = dis.readUTF();
        type = dis.readInt();

        id    = dis.readInt();
        refer = dis.readInt();
    }

    @NonNull
    @Override
    public String toString() {
        String format;
        if (refer > 0) format = String.format("ID:%d", id);
        else format = String.format("type[%d]： %d", type, text);
        return format;
    }
}

File: app/src/main/java/com/example/review/New/LibraryStruct.kt
package com.example.review.New

import com.example.review.DataStructureFile.ReviewData
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

class LibraryStruct : StoreData {
    var text = ""
        set(text) = if (refer > 0) field = "" else field = text
        get() {
            if (refer > 0) {
                for (ls in ReviewData.mLibraries) if (ls.id == refer) return ls.text
                return "**引用丢失！**"
            }
            return field
        }

    var type = 0
        get
        set
    var refer = 0
    var id = 0

    constructor() {}
    constructor(ls: LibraryStruct) {
        text = ls.text
        type = ls.type
    }

    fun copyOf(ls: LibraryStruct) { //是引用的话，则不能拷贝
        if (refer > 0) return
        //拷贝数据
        text = ls.text
        type = ls.type
    }

    constructor(text: String, type: Int) {
        this.text = text
        this.type = type
    }

    constructor(rawBytes: ByteArray) {
        try {
            loadWith(rawBytes)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun setIdAuto(value: Int) {
        if (ReviewData.Companion.mLibraries != null) {
            var maxId = 0
            for (ls in ReviewData.Companion.mLibraries) if (ls.id > maxId) maxId = ls.id
            id = maxId + 1 + value
        }
    }

    fun setIdAuto(library: LibraryList) {
        var maxId = 0
        for (ls in library) if (ls.id > maxId) maxId = ls.id
        id = maxId + 1
    }

    @Throws(IOException::class)
    override fun toBytes(dos: DataOutputStream) {
        dos.writeUTF(text)
        dos.writeInt(type)
        dos.writeInt(id)
        dos.writeInt(refer)
    }

    @Throws(IOException::class)
    override fun loadWith(dis: DataInputStream) {
        text = dis.readUTF()
        type = dis.readInt()
        id = dis.readInt()
        refer = dis.readInt()
    }

    override fun toString(): String = if (refer > 0) String.format("ID:%d", id) else String.format("type[%d]： %d", type, text)
}